src/flow_graph.py

```diff
@@ -1095,16 +1260,218 @@ def reset_block_info(self) -> None:
+        if isinstance(node, TerminalNode):
+            assert not node.block.instruction_refs
+            for reg in arch.all_return_regs:
+                add_source_uses("epilogue", reg)
+
+        # Process everything dominated by this node, now that we know our own
+        # register sources. This will eventually reach every node.
+        for child in node.immediately_dominates:
+            child_loc_srcs = loc_srcs.copy()
+
+            for loc in locs_clobbered_until_dominator(child):
+                phi_reg_srcs = phi_loc_sources(child, loc, loc_srcs.get(loc))
+                if phi_reg_srcs:
+                    # If phi_reg_srcs is non-empty, then loc is set in every control flow path
+                    # to the child node (but it may have a phi value)
+                    child_loc_srcs[loc] = phi_reg_srcs
+                elif loc in child_loc_srcs:
+                    # Otherwise, it's not always set, and cannot be used by the child node
+                    child_loc_srcs.remove(loc)
+
+            process_node(child, child_loc_srcs)
+
+    # Set all the registers that are valid to access at the start of a function
+    entry_reg_srcs = LocationRefSetDict()
+    for r in arch.all_regs:
+        entry_reg_srcs.refs[r] = RefSet.special("prologue")
+
+    # Recursively traverse every node, starting with the entry node, populating instr_inputs
+    entry_node = flow_graph.entry_node()
+    process_node(entry_node, entry_reg_srcs)
+
+    # Populate instr_uses for each instruction
+    for ref, inputs in flow_graph.instr_inputs.items():
+        for reg, deps in inputs.items():
+            for dep in deps:
+                if isinstance(dep, InstrRef):
+                    flow_graph.instr_uses[dep].add(reg, ref)
+
+    if print_warnings and missing_regs:
+        print("/*")
+        print(f"Warning: in {function.name}, regs were read before being written to:")
+        for reg, ref in missing_regs:
+            print(f"   {reg} at {ref}: {ref.instruction}")
+        print(f"*/")
+
+    return flow_graph
+
 
 def build_flowgraph(
-    function: Function, asm_data: AsmData, arch: ArchFlowGraph
+    function: Function,
+    asm_data: AsmData,
+    arch: ArchFlowGraph,
+    *,
+    fragment: bool,
+    print_warnings: bool = False,
 ) -> FlowGraph:
-    blocks = build_blocks(function, asm_data, arch)
+    blocks = build_blocks(function, asm_data, arch, fragment=fragment)
     nodes = build_nodes(function, blocks, asm_data, arch)
-    nodes = duplicate_premature_returns(nodes)
+    if not fragment:
+        nodes = duplicate_premature_returns(nodes)
+
     compute_relations(nodes)
-    terminate_infinite_loops(nodes)
-    return FlowGraph(nodes)
+    if not fragment:
+        terminate_infinite_loops(nodes)
+
+    flow_graph = nodes_to_flowgraph(
+        nodes, function, arch, print_warnings=print_warnings
+    )
+    if not fragment:
+        arch.simplify_ir(flow_graph)
+
+    return flow_graph
```

Comment:
(pondering whether it would be wise to create a separate build_fragment_flowgraph function to get rid of the "if not fragment"s, not sure)